Userland/Libraries/LibWeb/Crypto/SubtleCrypto.cpp

```diff
@@ -245,4 +344,49 @@ void SubtleCrypto::define_an_algorithm(String op, String algorithm, String type)
     internal_object.set(op, registered_algorithms);
 }
 
+// https://w3c.github.io/webcrypto/#pbkdf2-operations
+JS::ThrowCompletionOr<JS::NonnullGCPtr<CryptoKey>> SubtleCrypto::pbkdf2_import_key([[maybe_unused]] Variant<ByteBuffer, Bindings::JsonWebKey, Empty> key_data, [[maybe_unused]] AlgorithmIdentifier algorithm_parameter, Bindings::KeyFormat format, bool extractable, Vector<Bindings::KeyUsage> key_usages)
+{
+    auto& realm = this->realm();
+
+    // 1. If format is not "raw", throw a NotSupportedError
+    if (format != Bindings::KeyFormat::Raw) {
+        // FIXME: This should be a NotSupportedError
+        return realm.vm().throw_completion<JS::TypeError>(JS::ErrorType::NotImplemented, idl_enum_to_string(format));
+    }
+
+    // 2. If usages contains a value that is not "deriveKey" or "deriveBits", then throw a SyntaxError.
+    for (auto& usage : key_usages) {
+        if (usage != Bindings::KeyUsage::Derivekey && usage != Bindings::KeyUsage::Derivebits) {
+            return realm.vm().throw_completion<JS::SyntaxError>(MUST(String::formatted("Invalid key usage '{}'", idl_enum_to_string(usage))));
+        }
+    }
+
+    // 3. If extractable is not false, then throw a SyntaxError.
+    if (extractable) {
+        return realm.vm().throw_completion<JS::SyntaxError>("extractable must be false"_string);
+    }
+
+    // 4. Let key be a new CryptoKey representing keyData.
+    auto key = CryptoKey::create(realm);
+
+    // 5. Set the [[type]] internal slot of key to "secret".
+    key->set_type(Bindings::KeyType::Secret);
+
+    // 6. Set the [[extractable]] internal slot of key to false.
+    key->set_extractable(false);
+
+    // 7. Let algorithm be a new KeyAlgorithm object.
+    auto algorithm = Bindings::KeyAlgorithm::create(realm);
+
+    // 8. Set the name attribute of algorithm to "PBKDF2".
+    algorithm->set_name("PBKDF2"_string);
+
+    // 9. Set the [[algorithm]] internal slot of key to algorithm.
+    key->set_algorithm(algorithm);
```

Comment:
... I see. 

This is a super weird of dictionaries-as-records.

I would rather model these guys as raw POD structs with all the "slots" that the spec ever tries to dump values into.

If we want the key Algorithm to be GC allocated, then it should probably only derive from JS::Cell.

The spec also says that "The contents of the [[[algorithm]]](https://w3c.github.io/webcrypto/#dfn-CryptoKey-slot-algorithm) internal slot shall be, or be derived from, a [KeyAlgorithm](https://w3c.github.io/webcrypto/#dfn-KeyAlgorithm)."

Which is super great. But also that the actual js-visible `algorithm` property of the key is simply a `cached ECMAScript value derived from the [[algorithm]] internal slot` <https://w3c.github.io/webcrypto/#ref-for-dom-cryptokey-algorithm-1>

... does it actually say *how* to perform that derivation?

Reply:
.. according to mdn, for your first implementation here, the "algorithm_cached" internal slot that's actually returned via a "cached ecmascript value" (https://w3c.github.io/webcrypto/#concept-cached-object) for the "algorithm" property should be an object that models one of the KeyGenParams dictionaries?

https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey/algorithm#value

... otherwise it just needs to be "An object with a name property" ... which I guess matches what you have.